tests/unit/test_hnsw.cpp

```diff
@@ -1838,6 +1840,198 @@ TYPED_TEST(HNSWTest, rangeQueryCosine) {
     VecSimIndex_Free(index);
 }
 
+template <typename index_type_t>
+class HNSWSerializerTest : public HNSWTest<index_type_t> {
+protected:
+    ~HNSWSerializerTest() { remove(file_name.c_str()); }
+
+    void GenerateFileName(std::string base_name) {
+        file_name = base_name + VecSimType_ToString(index_type_t::get_index_type());
+    }
+
+    std::string file_name;
+};
+
+TYPED_TEST_SUITE(HNSWSerializerTest, DataTypeSet);
+TYPED_TEST(HNSWSerializerTest, hnsw_serialization_v2_info) {
+    size_t dim = 4;
+    size_t n = 1000;
+    size_t M = 8;
+    size_t ef = 8;
+
+    HNSWParams params{.dim = dim,
+                      .metric = VecSimMetric_L2,
+                      .initialCapacity = n,
+                      .blockSize = 1,
+                      .M = M,
+                      .efConstruction = ef,
+                      .efRuntime = ef};
+    VecSimIndex *index = this->CreateNewIndex(params);
+    HNSWIndex<TEST_DATA_T, TEST_DIST_T> *hnsw_index = this->CastToHNSW(index);
+
+    this->GenerateFileName(std::string(getenv("ROOT")) +
+                           "/tests/unit/data/1k-d4-L2-M8-ef8_info.hnsw_v2");
+    // Save index.
+    EXPECT_THROW(hnsw_index->saveIndex(this->file_name, Serializer::EncodingVersion_NOT_VALID),
+                 std::runtime_error);
+    hnsw_index->saveIndex(this->file_name, Serializer::EncodingVersion_V2);
+
+    // Get index info and copy it, so it will be available after the index is freed.
+    VecSimIndexInfo serialized_index_info = VecSimIndex_Info(index);
+    size_t serialized_allocation_size = hnsw_index->getAllocationSize();
+    // Free the index and initialize a new one with different parameters.
+    VecSimIndex_Free(index);
+
+    // bs, M, ef_c and ef_rt will be set to default
+    HNSWParams other_params{
+        .dim = dim / 2, .metric = VecSimMetric_Cosine, .initialCapacity = n / 2, .epsilon = 0.004};
+
+    index = this->CreateNewIndex(other_params);
+
+    VecSimIndexInfo other_index_info = VecSimIndex_Info(index);
+    // Make sure that this index is different
+    ASSERT_NE(serialized_index_info.hnswInfo.dim, other_index_info.hnswInfo.dim);
+    ASSERT_NE(serialized_index_info.hnswInfo.blockSize, other_index_info.hnswInfo.blockSize);
+    ASSERT_NE(serialized_index_info.hnswInfo.M, other_index_info.hnswInfo.M);
+    ASSERT_NE(serialized_index_info.hnswInfo.efConstruction,
+              other_index_info.hnswInfo.efConstruction);
+    ASSERT_NE(serialized_index_info.hnswInfo.efRuntime, other_index_info.hnswInfo.efRuntime);
+    ASSERT_NE(serialized_index_info.hnswInfo.epsilon, other_index_info.hnswInfo.epsilon);
+
+    hnsw_index = this->CastToHNSW(index);
+    hnsw_index->loadIndex(this->file_name);
```

Comment:
Looking at this test makes me think that we should use the `loadIndex` as an alternative constructor. Creating an index and then loading it from file was the temporary solution at the time, but now it looks awkward to use `loadIndex` to override index data. Let's discuss it...